packages/core/test/model.test.ts

```diff
@@ -72,6 +72,60 @@ describe('createModel', () => {
     expect(updatedState.context.name).toEqual('Anyone');
   });
 
+  it('model.assign should work as externally-defined action that does not restrict events', () => {
+    type UserEvent =
+      | {
+          type: 'updateName';
+          value: string;
+        }
+      | { type: 'updateAge'; value: number }
+      | { type: 'anotherEvent' };
+
+    interface UserContext {
+      name: string;
+      age: number;
+    }
+
+    const userModel = createModel<UserContext, UserEvent>({
+      name: 'David',
+      age: 30
+    });
+
+    userModel.assign({
+      name: (_, event) => {
+        // Incompatible because event.value can be string | number | undefined
+        // @ts-expect-error
+        return event.value;
```

Comment:
I think this is somewhat checking an incorrect thing here. It looks like your intention is to check assignability to the return value but actually you are suppressing property access error here:
https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzhATgUQG4QHbCgXgFBSFQA+UA3gUVaJAFxQDkArmACYCGwEAcuwLYQGAbkpVCqdgBsmEenGCIAlhgDmIqgF9RJclBqzGLDlwCCKwUKgTpBjEz4AjJFC1VSZPeAMN2GAPbAABZIaJjADC4ieABmTBgAxsCKfhhQ7HBwiioYABQQ6Fj08CEFwACU5NqIEMBMiKn5YQB01jIiGkA

This is because your `anotherEvent` doesn't declare `value` property at all - so you can't just freely access this property here without refining stuff first.

To fix that you could add `value?: undefined`  to the `anotherEvent`